src/components/ConfirmWithAuthModal.tsx

```diff
@@ -0,0 +1,122 @@
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { walletOpenAsyncUnsafe } from '@alephium/sdk'
+import { useEffect, useState } from 'react'
+import { Alert } from 'react-native'
+import styled from 'styled-components/native'
+
+import { errorInstructionSet, pinLength } from '../screens/LoginScreen'
+import { getStoredActiveWallet } from '../storage/wallets'
+import { ActiveWalletState } from '../store/activeWalletSlice'
+import { mnemonicToSeed, pbkdf2 } from '../utils/crypto'
+import PinCodeInput from './inputs/PinCodeInput'
+import ModalWithBackdrop from './ModalWithBackdrop'
+import SpinnerModal from './SpinnerModal'
+import CenteredInstructions, { Instruction } from './text/CenteredInstructions'
+
+interface ConfirmWithAuthModalProps {
+  onConfirm: () => void
+  onCancel: () => void
+}
+
+const firstInstructionSet: Instruction[] = [{ text: 'Please enter your pin', type: 'primary' }]
+
+const ConfirmWithAuthModal = ({ onConfirm, onCancel }: ConfirmWithAuthModalProps) => {
+  const [pinCode, setPinCode] = useState('')
+  const [shownInstructions, setShownInstructions] = useState(firstInstructionSet)
+  const [activeWallet, setActiveWallet] = useState<ActiveWalletState>()
+  const [requiresPin, setRequiresPin] = useState(false)
+  const [loading, setLoading] = useState(false)
+
+  useEffect(() => {
+    const getWallet = async () => {
+      try {
+        const storedActiveWallet = await getStoredActiveWallet()
+
+        if (storedActiveWallet) {
+          setActiveWallet(storedActiveWallet)
+
+          if (storedActiveWallet.authType === 'biometrics') {
+            onConfirm()
+          } else if (storedActiveWallet.authType === 'pin') {
+            setRequiresPin(true)
+          }
+        }
+      } catch (e: unknown) {
+        const error = e as { message?: string }
+
+        if (error.message === 'User canceled the authentication') {
+          onCancel()
+        } else if (error.message === 'No biometrics are currently enrolled') {
+          Alert.alert(
+            'Authentication required',
+            'To authenticate, please set up biometrics (fingerprint) on your device settings and try again.'
+          )
+        } else {
+          console.error(e)
+        }
+      }
+    }
+
+    getWallet()
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
```

Comment:
I want this effect to only run exactly 1 time. If I enable the exhaustive deps rule, I would need to add `onCancel` and `onConfirm` in the dependency list. Those 2 callbacks are defined here:

```ts
// src/screens/SendScreen.tsx
<ConfirmWithAuthModal onCancel={() => setIsAuthenticationModalVisible(false)} onConfirm={sendTransaction} />
```
https://github.com/alephium/mobile-wallet/blob/send-screen-additions/src/screens/SendScreen.tsx#L449

Since the `onCancel` callback gets re-defined every time `SendScreen` gets re-rendered, my effect would run **more than exactly once**, which is not desired. To avoid this I would have to extract the callback definition from the render function and wrap it around a `useCallback`. That would work for `onCancel` (although I find it overengineering for such a simple callback).

But it wouldn't for `onConfirm` because in the dependency list of the `sendTransaction`'s useCallback there is `fromAddress` which changes when we send a transaction (due to the pending tx added to it).  We could of course update the code to remove `fromAddress` from the depedency list, but that would require changing our `Client` class functions to accept `publicKey`, `privateKey`, and `hash` instead of just accepting the `Address` object.

All in all, even though I understand the need for this rule and even though it has saved my ass many times, I don't think it should be a hard rule.

I wonder what your opinion is on this.

Reply:
This is basically the diff. As you can see, the `onCancel` had to be extracted and wrapped in a `useCallback` and the dependency array of `sendTransaction` has to be tightened up (removing `fromAddress` and adding `fromAddress?.hash` and `fromAddress?.privateKey` instead, fields that do not change when adding a pending transaction to `fromAddress`.

Even though the code doesn't feel as clean now, I tend to feel that the code is more secure and more performant, since it causes fewer re-renders.

I look forward to your opinion on this.

<details><summary>Click to see it</summary>

```diff
diff --git a/src/api/client.ts b/src/api/client.ts
index 1c6e995..e0b3b32 100644
--- a/src/api/client.ts
+++ b/src/api/client.ts
@@ -38,8 +38,8 @@ export class Client {
     await this.cliqueClient.init(isMultiNodesClique)
   }
 
-  async buildSweepTransactions(address: Address, toHash: AddressHash) {
-    const { data } = await this.cliqueClient.transactionConsolidateUTXOs(address.publicKey, address.hash, toHash)
+  async buildSweepTransactions(fromHash: AddressHash, fromPublicKey: string, toHash: AddressHash) {
+    const { data } = await this.cliqueClient.transactionConsolidateUTXOs(fromPublicKey, fromHash, toHash)
     const fees = data.unsignedTxs.reduce((acc, tx) => acc + BigInt(tx.gasPrice) * BigInt(tx.gasAmount), BigInt(0))
 
     return {
@@ -48,9 +48,9 @@ export class Client {
     }
   }
 
-  async signAndSendTransaction(fromAddress: Address, txId: string, unsignedTx: string) {
-    const signature = this.cliqueClient.transactionSign(txId, fromAddress.privateKey)
-    const { data } = await this.cliqueClient.transactionSend(fromAddress.hash, unsignedTx, signature)
+  async signAndSendTransaction(fromHash: AddressHash, fromPrivateKey: string, txId: string, unsignedTx: string) {
+    const signature = this.cliqueClient.transactionSign(txId, fromPrivateKey)
+    const { data } = await this.cliqueClient.transactionSend(fromHash, unsignedTx, signature)
 
     return data
   }
diff --git a/src/components/ConfirmWithAuthModal.tsx b/src/components/ConfirmWithAuthModal.tsx
index 020ebbb..be6c025 100644
--- a/src/components/ConfirmWithAuthModal.tsx
+++ b/src/components/ConfirmWithAuthModal.tsx
@@ -75,8 +75,7 @@ const ConfirmWithAuthModal = ({ onConfirm, onCancel }: ConfirmWithAuthModalProps
     }
 
     getWallet()
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [])
+  }, [onCancel, onConfirm])
 
   useEffect(() => {
     if (!requiresPin || !pinCode || !activeWallet) return
diff --git a/src/screens/SendScreen.tsx b/src/screens/SendScreen.tsx
index cd00dec..76b73fd 100644
--- a/src/screens/SendScreen.tsx
+++ b/src/screens/SendScreen.tsx
@@ -136,7 +136,11 @@ const SendScreen = ({
     setIsLoading(true)
 
     try {
-      const { unsignedTxs, fees } = await client.buildSweepTransactions(fromAddress, fromAddress.hash)
+      const { unsignedTxs, fees } = await client.buildSweepTransactions(
+        fromAddress.hash,
+        fromAddress.publicKey,
+        fromAddress.hash
+      )
       setSweepUnsignedTxs(unsignedTxs)
       setFees(fees)
     } catch (e) {
@@ -148,7 +152,7 @@ const SendScreen = ({
 
   const buildTransaction = useCallback(
     async (formData: FormData) => {
-      if (!fromAddress || !isFormValid) return
+      if (!fromAddress?.hash || !isFormValid) return
 
       setIsLoading(true)
       const amountInSet = convertAlphToSet(formData.amountInAlph)
@@ -160,7 +164,11 @@ const SendScreen = ({
       const gasPriceInSet = formData.gasPriceInAlph ? convertAlphToSet(formData.gasPriceInAlph) : ''
       try {
         if (isSweep) {
-          const { unsignedTxs, fees } = await client.buildSweepTransactions(fromAddress, formData.toAddressHash)
+          const { unsignedTxs, fees } = await client.buildSweepTransactions(
+            fromAddress.hash,
+            fromAddress.publicKey,
+            formData.toAddressHash
+          )
           setSweepUnsignedTxs(unsignedTxs)
           setFees(fees)
         } else {
@@ -194,18 +202,24 @@ const SendScreen = ({
         setIsLoading(false)
       }
     },
-    [buildConsolidationTransactions, fromAddress, isFormValid]
+    [
+      buildConsolidationTransactions,
+      fromAddress?.hash,
+      fromAddress?.networkData.availableBalance,
+      fromAddress?.publicKey,
+      isFormValid
+    ]
   )
 
   const sendTransaction = useCallback(async () => {
-    if (!fromAddress) return
+    if (!fromAddress?.hash) return
 
     setIsLoading(true)
 
     try {
       if (isSweeping) {
         for (const { txId, unsignedTx } of sweepUnsignedTxs) {
-          client.signAndSendTransaction(fromAddress, txId, unsignedTx)
+          client.signAndSendTransaction(fromAddress.hash, fromAddress.privateKey, txId, unsignedTx)
 
           dispatch(
             addPendingTransactionToAddress({
@@ -219,7 +233,7 @@ const SendScreen = ({
           )
         }
       } else {
-        client.signAndSendTransaction(fromAddress, unsignedTxId, unsignedTransaction)
+        client.signAndSendTransaction(fromAddress.hash, fromAddress.privateKey, unsignedTxId, unsignedTransaction)
 
         dispatch(
           addPendingTransactionToAddress({
@@ -241,7 +255,8 @@ const SendScreen = ({
     amount,
     consolidationRequired,
     dispatch,
-    fromAddress,
+    fromAddress?.hash,
+    fromAddress?.privateKey,
     fromAddressHash,
     isSweeping,
     navigation,
@@ -265,6 +280,8 @@ const SendScreen = ({
     setValue('amountInAlph', convertSetToAlph(BigInt(fromAddress.networkData.availableBalance)))
   }, [fromAddress, setValue])
 
+  const handleAuthCancel = useCallback(() => setIsAuthenticationModalVisible(false), [])
+
   return (
     <Screen>
       <ScrollView>
@@ -446,7 +463,7 @@ const SendScreen = ({
           </ConsolidationModalContent>
         </ModalWithBackdrop>
         {isAuthenticationModalVisible && (
-          <ConfirmWithAuthModal onCancel={() => setIsAuthenticationModalVisible(false)} onConfirm={sendTransaction} />
+          <ConfirmWithAuthModal onCancel={handleAuthCancel} onConfirm={sendTransaction} />
         )}
       </ScrollView>
     </Screen>

```

</details>

Reply:
I decided to add a commit that fixes this, I can always revert it if needed.

Reply:
I've been wondering the same many times. I've encountered this problem many times, and most of those times I was able to use either `useCallback`, `useMemo` or code restructuration to avoid unwanted executions...
But sometimes it's indeed a huge pain in the ass...

So I'd say:
1) We try to make it work with one of the method cited above
2) But if we see that this is forcing us to overengineer too many things, we ignore the rule (if we're sure that it won't produce any unwanted side-effects).